sympy/integrals/tests/test_integrals.py

```diff
@@ -1,25 +1,63 @@
 from sympy import (
     Abs, acos, acosh, Add, And, asin, asinh, atan, Ci, cos, sinh, cosh,
-    tanh, Derivative, diff, DiracDelta, E, Ei, Eq, exp, erf, erfi,
+    tanh, Derivative, diff, DiracDelta, Dummy, E, Ei, Eq, exp, erf, erfi,
     EulerGamma, Expr, factor, Function, I, im, Integral, integrate,
-    Interval, Lambda, LambertW, log, Matrix, Max, meijerg, Min, nan,
+    Interval, Lambda, LambertW, limit, Limit, log, Matrix, Max, meijerg, Min, nan,
     Ne, O, oo, pi, Piecewise, polar_lift, Poly, Rational, re, S, Si, sign,
     simplify, sin, sinc, SingularityFunction, sqrt, sstr, Sum, Symbol,
     symbols, sympify, tan, trigsimp, Tuple
 )
 from sympy.functions.elementary.complexes import periodic_argument
 from sympy.functions.elementary.integers import floor
 from sympy.integrals.risch import NonElementaryIntegral
+from sympy.integrals.integrals import Integral
 from sympy.physics import units
 from sympy.core.compatibility import range
 from sympy.utilities.pytest import XFAIL, raises, slow, skip, ON_TRAVIS
 from sympy.utilities.randtest import verify_numerically
 
 
+
 x, y, a, t, x_1, x_2, z, s = symbols('x y a t x_1 x_2 z s')
 n = Symbol('n', integer=True)
 f = Function('f')
 
+def test_principal_value():
+
+    g = 1 / x
+    assert Integral(g, (x, -oo, oo)).principal_value() == 0
+    assert Integral(g, (y, -oo, oo)).principal_value() == oo*sign(1/x)
+    raises(ValueError, lambda: Integral(g, (x)).principal_value())
+    raises(ValueError, lambda: Integral(g).principal_value())
+
+    l = 1 / ((x**3) - 1)
+    assert Integral(l, (x, -oo, oo)).principal_value() == -sqrt(3)*pi/3
+    assert Integral(l, (x, -oo, 1)).principal_value() == -oo - I*pi/3
+    assert Integral(l, (x, -3, -2)).principal_value() == \
+           -sqrt(3)*atan(5*sqrt(3)/3)/3 - 2*log(2)/3 + log(3)/6 + log(7)/6 + sqrt(3)*pi/9
+    assert Integral(l, (x, 2, 3)).principal_value() == \
+           -sqrt(3) * atan(7 * sqrt(3) / 3) / 3 - log(13) / 6 + log(2) / 3 + log(7) / 6 + \
+           sqrt(3) * atan(5 * sqrt(3) / 3) / 3
+    raises(ValueError, lambda: Integral(l, (x)).principal_value())
+    raises(ValueError, lambda: Integral(l).principal_value())
+
+    k = 1 / (x - 4) * (x - 1)
+    assert Integral(k, (x, -oo, -2)).principal_value() == oo + 3*I*pi
+    assert Integral(k, (x, 5, oo)).principal_value() == oo
+    assert Integral(k, (x, 4, oo)).principal_value() == oo
+    raises(ValueError, lambda: Integral(k, (x)).principal_value())
+    raises(ValueError, lambda: Integral(k).principal_value())
+
+    d = (x ** 5) * (x - 1) / ((x ** 2 - 1) * (1 + x ** 2))
```

Comment:
It would be more useful to test functions with finite principal values. The numerator should be of lower degree.

There should be tests with many simpler functions like `1/(x**2 - 1)` or `1/((x ** 2 - 1) * (1 + x ** 2))` instead of a single function over many different intervals.

Reply:
@jksuom  Okay, Sir. Implementing..